src/event.c

```diff
@@ -314,23 +381,53 @@ bool event_loop(void) {
 	running = true;
 
 #ifndef HAVE_MINGW
+
+#ifdef HAVE_SYS_EPOLL_H
+
+	if(!epollset) {
+		epollset = event_epoll_init();
+	}
+
+#else
 	fd_set readable;
 	fd_set writable;
+#endif
 
 	while(running) {
 		struct timeval diff;
-		struct timeval *tv = get_time_remaining(&diff);
+		struct timeval *tv = timeout_execute(&diff);
+#ifndef HAVE_SYS_EPOLL_H
 		memcpy(&readable, &readfds, sizeof(readable));
 		memcpy(&writable, &writefds, sizeof(writable));
+#endif
+
+
+#ifdef HAVE_SYS_EPOLL_H
+		struct epoll_event events[EPOLL_MAX_EVENTS_PER_LOOP];
+		long timeout = (tv->tv_sec * 1000) + (tv->tv_usec / 1000);
 
-		int fds = 0;
+		if(timeout > INT_MAX) {
+			timeout = INT_MAX;
+		}
+
+		int n = epoll_wait(epollset, events, EPOLL_MAX_EVENTS_PER_LOOP, (int)timeout);
+#else
+		int maxfds = 1;
 
 		if(io_tree.tail) {
+			/*
+			        Use the max FD number to detirmine the max size of events
+			        This abuses the fact that:
+			        a) tinc does not allocate new sockets during operation
+			        b) that tinc does not have many sockets
+			        c) that sockets are allocated from 0 on all systems and so max(fd) is a reasonably approximation of nfds
```

Comment:
Yeah there were some experiments randomizing fds that turned out to upset a lot of programs. But the comments make it sound we do something bad, but we don't: at this point io_tree.tail really should point to the event with the highest fd number, no higher fd should be set in the fdsets. Nothing is abused.

Reply:
Biggest problems occur if tinc is linked with any shared library creating file descriptors for other purposes (e.g logging). Some comment explaining that you are using max(fd) to approximate nfds in the IO tree would be a good idea. But I'll remove it.

Reply:
Done

Reply:
Ah, maybe you misunderstood the purpose of the first argument to `select()`? The parameter name in the manpage (and the one tinc used) is perhaps confusing, it's not the number of filedescriptors that we have or are interested in, it's rather the size of the fdsets in bits that we care about. If it's set to 10, it will check the first 10 bits in all the fdsets starting from the first one, so it will check fds 0 to 9.